dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,243 @@
+from __future__ import unicode_literals
+
+import os
+import posixpath
+import logging
+
+from funcy import cached_property
+from backoff import on_exception, expo
+
+from dvc.scheme import Schemes
+from dvc.path_info import CloudURLInfo
+from dvc.remote.base import RemoteBASE
+from dvc.config import Config
+from dvc.exceptions import DvcException
+from dvc.remote.gdrive.pydrive import (
+    RequestListFile,
+    RequestListFilePaginated,
+    RequestCreateFolder,
+    RequestUploadFile,
+    RequestDownloadFile,
+)
+from dvc.remote.gdrive.utils import FOLDER_MIME_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+    GDRIVE_USER_CREDENTIALS_DATA = "GDRIVE_USER_CREDENTIALS_DATA"
+    CREDENTIALS_FILE_PATH = "credentials.json"
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
```

Comment:
Multiple pretty different and very related reasons:
- If user is not logged in to their Google Account, we should ask for login as soon as possible (it's just logical assumption that DVC shouldn't do any work if there is no access to remote drive).
- Multi-threaded nature of `Upload` (and maybe of others dvc API methods) forces us to initialize all common objects in the main thread to be shared between all parallel threads.
- Caching of directories' names and `Id`s inside remote root directory should be done to decrease number of Google API calls.

Reply:
@Suor can be resolved?

Reply:
The thing is we don't do that in other remotes, they are all connecting lazily. I don't understand what is happening here good enough honestly.

If we can't do anything about it then let's resolve it.

Reply:
I think it's still a good point. As far as I understand @Suor's concern, the way the code is written right now means that we potentially hit API (to cache ids) even if we don't actually use that remote after that. E.g. we parse configs and initialize remotes for every `dvc add` or something - I'm not sure about this. The point is that we'll have to be extra careful initializing remotes now or can even hit some unnecessary slow down.

I would try to make it lazy, there should be some simple `fancy` wrapper to do that I can imagine and @Suor could probably help with that.

Reply:
> E.g. we parse configs and initialize remotes for every dvc add or something - I'm not sure about this.

Remotes init seems doesn't happen on `dvc add`. Not sure what others potential commands with such issue might be.

> I would try to make it lazy, there should be some simple fancy wrapper to do that I can imagine and @Suor could probably help with that.

Initially, I've tried to have cached_property and do all `init_drive` logic inside cached_property (I think, it was inspired by available S3 logic or some other remote).
There appeared at least 2 issues:
- During parallelized files upload DVC tried to initialize cached_property for each thread.
- Something failed for data_cloud unit tests during tests init.

Will give it another try.

Reply:
I'm not sure that cached property is a way to go here. @Suor should know some good pattern how to make it lazy (and thread safe)

Reply:
I wonder why don't we ran into this in any other remote? Maybe because you can't create some dir/file twice and here we can. The thing is creating a dup file is an issue as it breaks the core supposition of dvc remotes that a path or path info identifies a remote object unambiguously. If some `scheme://bucket/path/to/file` will actually refer to 2 different instances this will break dvc. So, we should somehow protect against this. 

So do GDrive provide such functionality, or if not can we check for existence and then create without race? Simple thread lock won't unfortunately protect from concurrent dvc processes doing mess.

Reply:
It seems some time ago DVC had functionality to call `makedirs` before actual files uploading. It might be very suitable for GDrive to have missed folders creation stage again (or implement it).

Or we can create all 256 intermediate dirs in advance and check their existence on each DVC run. Creation of 256 dirs will take ~26 seconds.

> So do GDrive provide such functionality

Google API can reserve (return to us) range of `id`s which can be used later to create new folders (or files). That might be useful to get all folders ids in advance and "link" each unique id to unique folder name to avoid generation of different ids with similar folder names.
Complication is that this Google API is not exposed through PyDrive.
Last time there was discussion of this issue ( @efiop should remember ) and we agreed to allow GDrive to create multiple directories with similar titles (PyDrive queries should process directories with the same title as single directory). In theory, the possible max number of directories with the same title is the count of DVC's parallel jobs.

Reply:
The issue is not only directories. We might create a "file" several times in the same place.

Reply:
Do you mean with parallel push from multiple users at the same time? That's possible, but since the files will be similar to each other, does it break anything?

Reply:
If files are the same then it only breaks deduplication.

Reply:
I believe that files with the same path will have similar content, so it's duplication, but it's not critical for system operations.
btw, it seems there is no way to prevent creation of multiple folders/files for multiple simultaneous  `dvc push` processes into the same remote. It will require some kind of custom implemented synchronization at remote (`lock` file, etc).

Reply:
If remote is used for something else than cache then we don't have content defined by path and everything breaks.

Reply:
yep. This issue was considered by @efiop for the current implementation. The idea was if we can have some normalization script to remove duplicates. It seems `lock file` at remote was not considered as acceptable solution and having duplication looks as less evil in comparison with how hard it will be to have properly implemented `lock file` functionality

Reply:
So, let's first solve the lazy init problem? @Suor do we have a thread safe implementation in mind to do caching?

Reply:
> Caching of directories' names and Ids inside remote root directory should be done to decrease number of Google API calls.

Currently names and ids are retrieved from remote with single query and cached into dictionaries.

Reply:
How rclone solves the issue with duplicated files: https://rclone.org/drive/#duplicated-files

Reply:
`@cached_property` doesn't provide thread safetly, but we can do that manually:
```python
import threading
from funcy import threading

@property
@wrap_with(threading.Lock())
def gdrive(self):
    if not hasattr(self, '_gdrive'):
        # ... init code
        self._gdrive = client_instance

    return self._gdrive
```

I should probably add a `@threadsafe_cached_property` to funcy :).

Also, I don't think we should create all those dirs eagerly, this will make a big lag at the start even if we are uploading only a single file or if all 256 dirs are already created.

Reply:
`wrap_with(threading.RLock())` works good so far, thanks, resolving this conversation.